server/internal/infprocessor/infprocessor.go

```diff
@@ -293,16 +230,108 @@ func (p *P) findLeastLoadedEngine(engineIDs []string) string {
 	var minTasks int
 	var leastLoaded string
 	for _, engineID := range engineIDs {
+		if p.unavailableEngines[engineID] {
+			continue
+		}
 		n := numTasksByEngine[engineID]
 		if leastLoaded == "" || n < minTasks {
 			minTasks = n
 			leastLoaded = engineID
 		}
 	}
-
 	return leastLoaded
 }
 
+// SendChatCompletionTask sends a chat completion task.
+func (p *P) SendChatCompletionTask(
+	ctx context.Context,
+	tenantID string,
+	req *v1.CreateChatCompletionRequest,
+	header http.Header,
+) (*http.Response, error) {
+	return p.sendTask(ctx, tenantID, req, nil, header, p.logger.WithName("chat"))
+}
+
+// SendEmbeddingTask sends an embedding task.
+func (p *P) SendEmbeddingTask(
+	ctx context.Context,
+	tenantID string,
+	req *v1.CreateEmbeddingRequest,
+	header http.Header,
+) (*http.Response, error) {
+	return p.sendTask(ctx, tenantID, nil, req, header, p.logger.WithName("embedded"))
+}
+
+func (p *P) sendTask(
+	ctx context.Context,
+	tenantID string,
+	chatCompletionReq *v1.CreateChatCompletionRequest,
+	embeddingReq *v1.CreateEmbeddingRequest,
+	header http.Header,
+	logger logr.Logger,
+) (*http.Response, error) {
+	taskID, err := id.GenerateID("inf_", 24)
+	if err != nil {
+		return nil, fmt.Errorf("generate task ID: %s", err)
+	}
+	log := logger.WithValues("id", taskID)
+	done := make(chan *result, 1)
```

Comment:
Could you explain why the buffer size needs to be 1?

Reply:
This is because it guarantees the retrieval of one result and discards further results. The ordinary, done channel gets one result, but an additional error occurs when the engine is removed. Previously, the goroutine was used to send the error to the channel to prevent blocking in the removed engine. However, the goroutine would leak if the engine was removed during task processing second and later response since no one received the channel after getting the initial response.

Reply:
I see. Can you leave the comment? Thanks!